File: backend/scripts/replace-billing-data.py
#!/usr/bin/env python3
"""
Replace Billing Data Script (Python version)

Completely replaces data in billing-related tables with correct data
from desilope_softaware.sql, mapping both table names AND column names.

This Python version uses mysql-connector-python's batch execution
which is more robust for complex INSERT statements.
"""

import os
import re
import mysql.connector
from mysql.connector import Error

# Column mapping configurations
TABLE_CONFIGS = {
    'tb_contacts': {
        'new_table': 'contacts',
        'column_map': {
            'contact_id': 'id',
            'contact_name': 'company_name',
            'contact_person': 'contact_person',
            'contact_email': 'email',
            'contact_phone': 'phone',
            'contact_alt_phone': 'fax',
            'contact_address': 'location',
            'contact_notes': 'remarks',
            'contact_vat': None,  # Skip
            'contact_type': None  # Skip
        }
    },
    'tb_expense_categories': {
        'new_table': 'expense_categories',
        'direct':  True  # Just map table name, columns likely match
    },
    'tb_invoices': {
        'new_table': 'invoices',
        'direct': True
    },
    'tb_invoice_items': {
        'new_table': 'invoice_items',
        'direct': True
    },
    'tb_payments': {
        'new_table': 'payments',
        'direct': True
    },
    'tb_pricing': {
        'new_table': 'pricing',
        'direct': True
    },
    'tb_quotations': {
        'new_table': 'quotations',
        'direct': True
    },
    'tb_quote_items': {
        'new_table': 'quote_items',
        'direct': True
    },
    'tb_transactions': {
        'new_table': 'transactions',
        'direct': True
    }
}

def parse_db_url(url):
    """Parse DATABASE_URL format"""
    match = re.match(r'mysql://([^:]+):([^@]+)@([^:]+):(\d+)/(.+)', url)
    if not match:
        raise ValueError('Invalid DATABASE_URL format')
    return {
        'user': match.group(1),
        'password': match.group(2),
        'host': match.group(3),
        'port': int(match.group(4)),
        'database': match.group(5),
    }

def extract_inserts(sql_content, table_name):
    """Extract all INSERT statements for a table"""
    statements = []
    # Match INSERT statements - handling multi-line
    pattern = rf"INSERT INTO `{table_name}`[^;]+;"
    matches = re.finditer(pattern, sql_content, re.IGNORECASE | re.DOTALL)
    for match in matches:
        statements.append(match.group(0))
    return statements

def transform_insert_simple(old_table, new_table, column_map, statement):
    """Transform INSERT by replacing table name and filtering columns"""
    # First, just replace table name
    if column_map is None:  # Direct mapping
        return statement.replace(f'`{old_table}`', f'`{new_table}`')
    
    # Parse the INSERT structure
    # Match: INSERT INTO `table` (`col1`, `col2`) VALUES ...
    match = re.match(
        rf"INSERT INTO `{old_table}` \(([^)]+)\) VALUES\s*(.+);",
        statement,
        re.DOTALL
    )
    
    if not match:
        raise ValueError(f"Could not parse INSERT statement for {old_table}")
    
    columns_str = match.group(1)
    values_str = match.group(2)
    
    # Parse columns
    old_columns = [c.strip().replace('`', '') for c in columns_str.split(',')]
    
    # Build new columns list and indices to keep
    new_columns = []
    keep_indices = []
    
    for i, old_col in enumerate(old_columns):
        new_col = column_map.get(old_col)
        if new_col is not None:
            new_columns.append(new_col)
            keep_indices.append(i)
    
    if not new_columns:
        raise ValueError(f"No columns mapped for {old_table}")
    
    # Now we need to filter values
    # This is complex because values can contain commas in strings
    # For now, let's use a MySQL-compatible approach: just execute row by row
    return None, old_columns, new_columns, keep_indices, values_str

def get_table_schema(cursor, table_name):
    """Get table schema"""
    cursor.execute(f"DESCRIBE `{table_name}`")
    return [{'field': row[0], 'type': row[1]} for row in cursor.fetchall()]

def clear_table(cursor, table_name):
    """Clear table data"""
    print(f"  🗑️  Clearing {table_name}...")
    cursor.execute("SET FOREIGN_KEY_CHECKS = 0")
    cursor.execute(f"TRUNCATE TABLE `{table_name}`")
    cursor.execute("SET FOREIGN_KEY_CHECKS = 1")

def load_data_direct(cursor, statements, old_table, new_table):
    """Load data using direct table name replacement"""
    if not statements:
        print(f"  ⚠️  No data found")
        return 0
    
    print(f"  📥 Loading {len(statements)} INSERT statement(s)...")
    cursor.execute("SET FOREIGN_KEY_CHECKS = 0")
    
    total_rows = 0
    for stmt in statements:
        # Replace table name
        new_stmt = stmt.replace(f'`{old_table}`', f'`{new_table}`')
        try:
            cursor.execute(new_stmt)
            total_rows += cursor.rowcount
        except Error as e:
            print(f"  ❌ Error: {e}")
            print(f"  Statement preview: {new_stmt[:200]}...")
            raise
    
    cursor.execute("SET FOREIGN_KEY_CHECKS = 1")
    return total_rows

def verify_count(cursor, table_name):
    """Get row count"""
    cursor.execute(f"SELECT COUNT(*) FROM `{table_name}`")
    return cursor.fetchone()[0]

def main():
    print("🔄 Billing Data Replacement Script (Python)")
    print("=" * 60)
    print()
    
    # Read SQL dump
    print("📖 Step 1: Reading SQL dump...")
    sql_path = '/var/opt/backend/desilope_softaware.sql'
    
    with open(sql_path, 'r', encoding='utf-8') as f:
        sql_content = f.read()
    
    print(f"  ✓ Loaded {len(sql_content) / 1024 / 1024:.2f} MB")
    print()
    
    # Connect to database
    print("🔌 Step 2: Connecting to database...")
    db_url = os.getenv('DATABASE_URL')
    if not db_url:
        raise ValueError('DATABASE_URL not set')
    
    db_config = parse_db_url(db_url)
    print(f"  Database: {db_config['database']}")
    print(f"  Host: {db_config['host']}:{db_config['port']}")
    
    conn = mysql.connector.connect(
        host=db_config['host'],
        port=db_config['port'],
        user=db_config['user'],
        password=db_config['password'],
        database=db_config['database'],
        autocommit=True
    )
    
    cursor = conn.cursor()
    print("  ✓ Connected")
    print()
    
    try:
        print("🔄 Step 3: Replacing table data...")
        print()
        
        results = {}
        
        for old_table, config in TABLE_CONFIGS.items():
            new_table = config['new_table']
            print(f"📋 Processing {old_table} → {new_table}...")
            
            # Extract INSERT statements
            statements = extract_inserts(sql_content, old_table)
            print(f"  📄 Found {len(statements)} INSERT statement(s)")
            
            if not statements:
                print(f"  ⚠️  No data found for {old_table}")
                print()
                continue
            
            # Clear existing data
            clear_table(cursor, new_table)
            
            # Load data
            if config.get('direct'):
                # Simple table name replacement
                rows = load_data_direct(cursor, statements, old_table, new_table)
            else:
                # Complex column mapping (for tb_contacts)
                print(f"  🔄 Using column mapping...")
                # For contacts, we need special handling
                # Let's skip contacts for now and do the others first
                print(f"  ⚠️  Skipping {old_table} - requires custom mapping")
                print()
                continue
            
            # Verify
            actual_count = verify_count(cursor, new_table)
            
            results[new_table] = {
                'old_name': old_table,
                'statements': len(statements),
                'actual_count': actual_count
            }
            
            print(f"  ✓ {actual_count} rows now in {new_table}")
            print()
        
        # Summary
        print("=" * 60)
        print("📊 REPLACEMENT SUMMARY")
        print("=" * 60)
        print()
        
        total_rows = 0
        for table, stats in results.items():
            print(f"{stats['old_name']:<25} → {table:<20} {stats['actual_count']:>6} rows")
            total_rows += stats['actual_count']
        
        print("-" * 60)
        print(f"{'TOTAL':<47} {total_rows:>6} rows")
        print()
        print("✅ Data replacement completed successfully!")
        print()
        print("⚠️  Note: tb_contacts was skipped and needs custom column mapping")
        
    except Exception as e:
        print()
        print(f"❌ Error during replacement: {e}")
        import traceback
        traceback.print_exc()
        raise
    finally:
        cursor.close()
        conn.close()

if __name__ == '__main__':
    main()

File: backend/scripts/replace-billing-data.mjs
#!/usr/bin/env node

/**
 * Replace Billing Data Script
 * 
 * Completely replaces data in billing-related tables with correct data
 * from desilope_softaware.sql, including IDs.
 * 
 * Tables to replace:
 * - tb_contacts
 * - tb_expense_categories
 * - tb_invoices
 * - tb_invoice_items
 * - tb_payments
 * - tb_pricing
 * - tb_quotations
 * - tb_quote_items
 * - tb_transactions
 * 
 * Usage:
 *   node scripts/replace-billing-data.mjs
 */

import fs from 'fs';
import mysql from 'mysql2/promise';
import dotenv from 'dotenv';

// Load environment variables
dotenv.config({ path: '/var/opt/backend/.env' });

// Map old table names (from PHP SQL dump) to new table names (SoftAware backend)
const TABLE_MAPPING = {
  'tb_contacts': 'contacts',
  'tb_expense_categories': 'expense_categories',
  'tb_invoices': 'invoices',
  'tb_invoice_items': 'invoice_items',
  'tb_payments': 'payments',
  'tb_pricing': 'pricing',
  'tb_quotations': 'quotations',
  'tb_quote_items': 'quote_items',
  'tb_transactions': 'transactions'
};

// Parse DATABASE_URL
function parseDbUrl(url) {
  const match = url.match(/^mysql:\/\/([^:]+):([^@]+)@([^:]+):(\d+)\/(.+)$/);
  if (!match) throw new Error('Invalid DATABASE_URL format');
  return {
    user: match[1],
    password: match[2],
    host: match[3],
    port: parseInt(match[4]),
    database: match[5],
  };
}

/**
 * Extract CREATE TABLE statement for a specific table
 */
function extractCreateStatement(sqlContent, tableName) {
  // Find the CREATE TABLE statement
  const createRegex = new RegExp(
    `CREATE TABLE \`${tableName}\`[^;]+;`,
    'si'
  );
  const match = sqlContent.match(createRegex);
  return match ? match[0] : null;
}

/**
 * Extract INSERT statements for a specific table
 */
function extractInsertStatements(sqlContent, tableName) {
  const statements = [];
  
  // Find all INSERT INTO statements for this table
  const regex = new RegExp(
    `INSERT INTO \`${tableName}\`[^;]+;`,
    'gi'
  );
  
  let match;
  while ((match = regex.exec(sqlContent)) !== null) {
    statements.push(match[0]);
  }
  
  return statements;
}

/**
 * Disable foreign key checks and truncate table
 */
async function clearTable(conn, tableName) {
  console.log(`  🗑️  Clearing ${tableName}...`);
  await conn.query('SET FOREIGN_KEY_CHECKS = 0');
  await conn.query(`TRUNCATE TABLE \`${tableName}\``);
  await conn.query('SET FOREIGN_KEY_CHECKS = 1');
}

/**
 * Load data from INSERT statements, replacing old table name with new table name
 */
async function loadTableData(conn, oldTableName, newTableName, insertStatements) {
  if (insertStatements.length === 0) {
    console.log(`  ⚠️  No data found for ${oldTableName}`);
    return 0;
  }
  
  console.log(`  📥 Loading ${insertStatements.length} INSERT statement(s) into ${newTableName}...`);
  
  let totalRows = 0;
  await conn.query('SET FOREIGN_KEY_CHECKS = 0');
  
  for (const statement of insertStatements) {
    try {
      // Replace old table name with new table name
      const modifiedStatement = statement.replace(
        new RegExp(`INSERT INTO \`${oldTableName}\``, 'gi'),
        `INSERT INTO \`${newTableName}\``
      );
      
      const [result] = await conn.query(modifiedStatement);
      const affectedRows = result.affectedRows || 0;
      totalRows += affectedRows;
    } catch (err) {
      console.error(`  ❌ Error executing INSERT for ${newTableName}:`, err.message);
      throw err;
    }
  }
  
  await conn.query('SET FOREIGN_KEY_CHECKS = 1');
  
  return totalRows;
}

/**
 * Verify row count
 */
async function verifyCount(conn, tableName) {
  const [rows] = await conn.query(`SELECT COUNT(*) as count FROM \`${tableName}\``);
  return rows[0].count;
}

async function main() {
  console.log('🔄 Billing Data Replacement Script');
  console.log('=' .repeat(60));
  console.log();
  
  // Read SQL dump
  console.log('📖 Step 1: Reading SQL dump...');
  const sqlPath = '/var/opt/backend/desilope_softaware.sql';
  
  if (!fs.existsSync(sqlPath)) {
    throw new Error(`SQL file not found: ${sqlPath}`);
  }
  
  const sqlContent = fs.readFileSync(sqlPath, 'utf-8');
  console.log(`  ✓ Loaded ${(sqlContent.length / 1024 / 1024).toFixed(2)} MB`);
  console.log();
  
  // Parse database configuration
  console.log('🔌 Step 2: Connecting to database...');
  const dbUrl = process.env.DATABASE_URL;
  if (!dbUrl) {
    throw new Error('DATABASE_URL not set in environment');
  }
  
  const dbConfig = parseDbUrl(dbUrl);
  console.log(`  Database: ${dbConfig.database}`);
  console.log(`  Host: ${dbConfig.host}:${dbConfig.port}`);
  
  const connection = await mysql.createConnection({
    host: dbConfig.host,
    port: dbConfig.port,
    user: dbConfig.user,
    password: dbConfig.password,
    database: dbConfig.database,
    multipleStatements: true,
  });
  
  console.log('  ✓ Connected');
  console.log();
  
  try {
    console.log('🔄 Step 3: Replacing table data...');
    console.log();
    
    const results = {};
    
    for (const [oldTableName, newTableName] of Object.entries(TABLE_MAPPING)) {
      console.log(`📋 Processing ${oldTableName} → ${newTableName}...`);
      
      // Extract INSERT statements from old table name
      const insertStatements = extractInsertStatements(sqlContent, oldTableName);
      
      // Clear existing data in new table
      await clearTable(connection, newTableName);
      
      // Load new data into new table
      const rowsInserted = await loadTableData(connection, oldTableName, newTableName, insertStatements);
      
      // Verify
      const actualCount = await verifyCount(connection, newTableName);
      
      results[newTableName] = {
        oldName: oldTableName,
        statements: insertStatements.length,
        rowsInserted,
        actualCount
      };
      
      console.log(`  ✓ ${actualCount} rows now in ${newTableName}`);
      console.log();
    }
    
    // Summary
    console.log('=' .repeat(60));
    console.log('📊 REPLACEMENT SUMMARY');
    console.log('=' .repeat(60));
    console.log();
    
    let totalRows = 0;
    for (const [table, stats] of Object.entries(results)) {
      console.log(`${table.padEnd(30)} ${stats.actualCount.toString().padStart(10)} rows`);
      totalRows += stats.actualCount;
    }
    
    console.log('-'.repeat(60));
    console.log(`${'TOTAL'.padEnd(30)} ${totalRows.toString().padStart(10)} rows`);
    console.log();
    console.log('✅ Data replacement completed successfully!');
    
  } catch (error) {
    console.error();
    console.error('❌ Error during replacement:');
    console.error(error);
    throw error;
  } finally {
    await connection.end();
  }
}

// Run the script
main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});
